File: 820-find-eventual-safe-states/find-eventual-safe-states.py
class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        reversedGraph = [[] for _ in range(len(graph))]
        inDegrees = [0 for _ in range(len(graph))]
        result = []
        for node in range(len(graph)):
            for nextNode in graph[node]:
                reversedGraph[nextNode].append(node)
                inDegrees[node]+=1
        # print(graph)
        # print(reversedGraph)
        # print(inDegrees)
        queue = deque()
        for i,degree in enumerate(inDegrees):
            if degree == 0:
                queue.append(i)
        # print(queue)
        while queue:
            currNode = queue.popleft()
            result.append(currNode)
            for adjNode in reversedGraph[currNode]:
                inDegrees[adjNode]-=1
                if inDegrees[adjNode] == 0:
                    queue.append(adjNode)
        return sorted(result)

File: 820-find-eventual-safe-states/find-eventual-safe-states.java
class Solution {
    public int [][] graph;
    public List<Integer> eventualSafeNodes(int[][] graph) {
        this.graph = graph;
        int [] visited = new int[graph.length];
        Arrays.fill(visited,-1);
        List<Integer> safe = new ArrayList<>();
        for(int i = 0;i<graph.length;i++){
            if(dfs(i,visited)){
                safe.add(i);
            }
        }
        return safe;
    }
    public boolean dfs(int node,int [] visited){
        if(visited[node] == 1){
            return false;
        }
        if(visited[node] == 2){
            return true;
        }
        visited[node] = 1;
        for(int i = 0;i<graph[node].length;i++){
            if(!dfs(graph[node][i],visited)){
                return false;
            }
        }
        visited[node] = 2;
        return true;
    }
}